### PR TITLE
cmd/scollector: Add os.net.bond.* and win.net.tcp.retransmit_pct

### PR DESCRIPTION
**Pull request by gbrayut**

### CODE REVIEW

File: cmd/scollector/collectors/collectors.go [hunk 1]
```diff
@@ -24,41 +24,57 @@ type Collector interface {
...
+	osMemTotal         = "os.mem.total"
+	osMemUsed          = "os.mem.used"
+	osNetBroadcast     = "os.net.packets_broadcast"
+	osNetBytes         = "os.net.bytes"
+	osNetDropped       = "os.net.dropped"
+	osNetErrors        = "os.net.errs"
+	osNetMulticast     = "os.net.packets_multicast"
+	osNetPackets       = "os.net.packets"
+	osNetUnicast       = "os.net.packets_unicast"
+	osNetBondBytes     = "os.net.bond.bytes"
```
Comment (mjibson) [inline cmd/scollector/collectors/collectors.go]: Could you alphabetize this list?

File: cmd/scollector/collectors/collectors.go [hunk 2]
```diff
@@ -24,41 +24,57 @@ type Collector interface {
...
+	osDiskUsedDesc         = "The space_used property indicates in bytes how much space is used on the disk."
+	osMemTotalDesc         = "Total amount, in bytes, of physical memory available to the operating system."
+	osMemFreeDesc          = "Number, in bytes, of physical memory currently unused and available."
+	osMemPctFreeDesc       = "The percent of free memory. In Linux free memory includes memory used by buffers and cache."
+	osMemUsedDesc          = "The amount of used memory. In Linux this excludes memory used by buffers and cache."
+	osNetBytesDesc         = "The rate at which bytes are sent or received over each network adapter."
+	osNetDroppedDesc       = "The number of packets that were chosen to be discarded even though no errors had been detected to prevent transmission."
+	osNetErrorsDesc        = "The number of packets that could not be transmitted because of errors."
+	osNetPacketsDesc       = "The rate at which packets are sent or received on the network interface."
+	osNetBondBytesDesc     = "The rate at which bytes are sent or received over bonded network adapters."
```
Comment (mjibson) [inline cmd/scollector/collectors/collectors.go]: Alphabetize

File: cmd/scollector/collectors/network_windows.go [hunk 3]
```diff
@@ -314,6 +333,13 @@ func c_network_windows_tcp() (opentsdb.MultiDataPoint, error) {
 		Add(&md, "win.net.tcp.segments", v.SegmentsReceivedPersec, opentsdb.TagSet{"type": "received"}, metadata.Counter, metadata.PerSecond, descWinNetTCPv4SegmentsReceivedPersec)
 		Add(&md, "win.net.tcp.segments", v.SegmentsRetransmittedPersec, opentsdb.TagSet{"type": "retransmitted"}, metadata.Counter, metadata.PerSecond, descWinNetTCPv4SegmentsRetransmittedPersec)
 		Add(&md, "win.net.tcp.segments", v.SegmentsSentPersec, opentsdb.TagSet{"type": "sent"}, metadata.Counter, metadata.PerSecond, descWinNetTCPv4SegmentsSentPersec)
+		if (v.SegmentsPersec-winNetTCPSegmentsLastCount) != 0 &&
+			(v.SegmentsRetransmittedPersec > winNetTCPSegmentsLastRetransmitCount) &&
+			(v.SegmentsPersec > winNetTCPSegmentsLastCount) {
+			Add(&md, "win.net.tcp.retransmit_pct", float64(v.SegmentsRetransmittedPersec-winNetTCPSegmentsLastRetransmitCount)/float64(v.SegmentsPersec-winNetTCPSegmentsLastCount)*100, nil, metadata.Gauge, metadata.Pct, descWinNetTCPv4SegmentsRetransmit)
```
Comment (mjibson) [inline cmd/scollector/collectors/network_windows.go]: This line is pretty long. How about:

```
val := float64(v.SegmentsRetransmittedPersec-winNetTCPSegmentsLastRetransmitCount)/float64(v.SegmentsPersec-winNetTCPSegmentsLastCount)*100
Add(&md, "win.net.tcp.retransmit_pct", val, nil, metadata.Gauge, metadata.Pct, descWinNetTCPv4SegmentsRetransmit)
```

File: cmd/scollector/collectors/network_windows.go [hunk 4]
```diff
@@ -239,7 +239,21 @@ func c_network_team_windows() (opentsdb.MultiDataPoint, error) {
...
+		Add(&md, osNetBondBytes, nicStats.SentBytes, tagsOut, metadata.Counter, metadata.Bytes, osNetBondBytesDesc)
+		Add(&md, osNetBondUnicast, nicStats.ReceivedUnicastPackets, tagsIn, metadata.Counter, metadata.Count, osNetBondUnicastDesc)
+		Add(&md, osNetBondUnicast, nicStats.SentUnicastPackets, tagsOut, metadata.Counter, metadata.Count, osNetBondUnicastDesc)
+		Add(&md, osNetBondMulticast, nicStats.ReceivedMulticastPackets, tagsIn, metadata.Counter, metadata.Count, osNetBondMulticastDesc)
+		Add(&md, osNetBondMulticast, nicStats.SentMulticastPackets, tagsOut, metadata.Counter, metadata.Count, osNetBondMulticastDesc)
+		Add(&md, osNetBondBroadcast, nicStats.ReceivedBroadcastPackets, tagsIn, metadata.Counter, metadata.Count, osNetBondBroadcastDesc)
+		Add(&md, osNetBondBroadcast, nicStats.SentBroadcastPackets, tagsOut, metadata.Counter, metadata.Count, osNetBondBroadcastDesc)
+		Add(&md, osNetBondPackets, float64(nicStats.ReceivedUnicastPackets)+float64(nicStats.ReceivedMulticastPackets)+float64(nicStats.ReceivedBroadcastPackets), tagsIn, metadata.Counter, metadata.Count, osNetBondPacketsDesc)
```
Comment (mjibson) [inline cmd/scollector/collectors/network_windows.go]: Why do these need to be converted to float64?

File: cmd/scollector/collectors/network_windows.go [hunk 5]
```diff
@@ -239,7 +239,21 @@ func c_network_team_windows() (opentsdb.MultiDataPoint, error) {
...
+		Add(&md, osNetBondPackets, float64(nicStats.SentUnicastPackets)+float64(nicStats.SentMulticastPackets)+float64(nicStats.SentBroadcastPackets), tagsOut, metadata.Counter, metadata.Count, osNetBondPacketsDesc)
```
Comment (mjibson) [inline cmd/scollector/collectors/network_windows.go]: Why do these need to be converted to float64?

File: cmd/scollector/collectors/network_windows.go [hunk 6]
```diff
@@ -314,6 +333,13 @@ func c_network_windows_tcp() (opentsdb.MultiDataPoint, error) {
 		Add(&md, "win.net.tcp.segments", v.SegmentsReceivedPersec, opentsdb.TagSet{"type": "received"}, metadata.Counter, metadata.PerSecond, descWinNetTCPv4SegmentsReceivedPersec)
 		Add(&md, "win.net.tcp.segments", v.SegmentsRetransmittedPersec, opentsdb.TagSet{"type": "retransmitted"}, metadata.Counter, metadata.PerSecond, descWinNetTCPv4SegmentsRetransmittedPersec)
 		Add(&md, "win.net.tcp.segments", v.SegmentsSentPersec, opentsdb.TagSet{"type": "sent"}, metadata.Counter, metadata.PerSecond, descWinNetTCPv4SegmentsSentPersec)
+		if (v.SegmentsPersec-winNetTCPSegmentsLastCount) != 0 &&
+			(v.SegmentsRetransmittedPersec > winNetTCPSegmentsLastRetransmitCount) &&
+			(v.SegmentsPersec > winNetTCPSegmentsLastCount) {
+			Add(&md, "win.net.tcp.retransmit_pct", float64(v.SegmentsRetransmittedPersec-winNetTCPSegmentsLastRetransmitCount)/float64(v.SegmentsPersec-winNetTCPSegmentsLastCount)*100, nil, metadata.Gauge, metadata.Pct, descWinNetTCPv4SegmentsRetransmit)
+		}
+		winNetTCPSegmentsLastRetransmitCount = v.SegmentsRetransmittedPersec
```
Comment (mjibson) [inline cmd/scollector/collectors/network_windows.go]: Why do we need to keep these around? I can't find another collector that maintains state like this.

Comment (gbrayut) [inline cmd/scollector/collectors/network_windows.go]: Need to turn a counter into a gauge. If I just use the raw values, I get the percent for all time, which on sql is misleading (the retransmit doesn't reset the counter as often as the others)

File: cmd/scollector/collectors/network_windows.go (hunk 4 above)
Comment (gbrayut) [inline cmd/scollector/collectors/network_windows.go]: Not sure if they do. See https://bosun.slack.com/archives/general/p1432234382000592 in bosun chat

**Comment by gbrayut** [comment]
Chatted with Matt... Keeping osNetBondPackets as float64s to prevent excessive overflowing after combining the 3 uint64s. Also added check to discard initial win.net.tcp.retransmit_pct datapoint.

Merging

**Pull request merged**